HDR Rendering

bagnell opened the pull request:
Adds the ability to render with high dynamic range.
* `Scene.highDynamicRange` will enable disable HDR rendering when supported. Defaults to `true`
* `Scene.gamma` will change the gamma correction. Defaults to `2.2`.

Auto-exposure was added but is hard-coded to be disabled. Most built-in primitives still render LDR values so the average luminance of the scene is normally < 1.0 which would make everything much brighter.

The sun, sky atmosphere, ground atmosphere, and potentially specular highlights from PBR models use the higher range.

Requires #6877 and #6943 to be merged first.

TODO:
- [x] Fix issue when toggling HDR (There is no issue when set before the first render).
- [x] Fix tests.
- [x] Add tests.
- [x] Update CHANGES.md

Future:
* Update all primitives to output HDR values.
* Enable auto-exposure
* Update auto exposure to [use linear sampling](https://github.com/AnalyticalGraphicsInc/cesium/pull/7017#discussion_r224792080)
* Add different bloom algorithm if floating-point texture linear interpolation is supported.
* Expose different tone mapping algorithms.
* Re-add lens flare not just when in space.

cesium-concierge commented:
Thanks for the pull request @bagnell!
* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

pjcozzi commented on Source/Renderer/AutomaticUniforms.js:
```diff
@@ -1686,6 +1686,34 @@ define([
+        }),
+
+        /**
+         * An automatic GLSL uniform that will be used for gamma correction.
```
"will be" is awkward, right?  "is?"

pjcozzi commented on Source/Renderer/AutomaticUniforms.js:
```diff
@@ -1686,6 +1686,34 @@ define([
+         */
+        czm_sunColor: new AutomaticUniform({
+            size: 1,
+            datatype: WebGLConstants.FLOAT_VEC4,
```
What is alpha used for?

pjcozzi commented on Source/Scene/AutoExposure.js:
```diff
@@ -0,0 +1,416 @@
+    'use strict';
+
+    /**
+     * A post process stage that will get the luminance value at each pixel and
```
> to get the average luminance

"to compute?"

pjcozzi commented on Source/Scene/AutoExposure.js:
```diff
@@ -0,0 +1,416 @@
+     * @constructor
+     * @private
+     */
+    function AutoExposure() {
```
I know this is private, but should it be `AutomaticExposure`?  Or is "AutoExposure" such a common abbreviation that we can allow it like WGS84?

pjcozzi commented on Source/Scene/AutoExposure.js:
```diff
@@ -0,0 +1,416 @@
+        this.enabled = true;
+        this._enabled = true;
+
+        this.minimumLuminance = 0.1;
```
Should here and below have doc since the others do?

pjcozzi commented on Source/Scene/AutoExposure.js:
```diff
@@ -0,0 +1,416 @@
+            };
+        }
+
+        uniforms.minimumLuminance = function() {
```
Can these three uniforms be combined or is that hard to do in the post-process framework?

pjcozzi commented on Source/Scene/AutoExposure.js:
```diff
@@ -0,0 +1,416 @@
+            'uniform sampler2D previousLuminance; \n' +
+            'void main() { \n' +
+            '    float color = 0.0; \n' +
+            '    float xStep = 1.0 /colorTextureDimensions.x; \n' +
```
Whitespace after `/` here and below.

pjcozzi commented on Source/Scene/AutoExposure.js:
```diff
@@ -0,0 +1,416 @@
+
+        source += '}\n\n';
+
+        source +=
```
The answer may be no, but it is possible and reasonable to pull the parallel reduction core into a standalone `czm_` function that we may use elsewhere, e.g., for edge-preserving blur?

pjcozzi commented on Source/Scene/AutoExposure.js:
```diff
@@ -0,0 +1,416 @@
+    AutoExposure.prototype.execute = function(context, colorTexture) {
+        this._colorTexture = colorTexture;
+
+        var commands = this._commands;
```
Do other post-processing stages or primitives have code like this to execute an array of commands?  If so, can we move it to a reusable function?

pjcozzi commented on Source/Scene/DerivedCommand.js:
```diff
@@ -323,5 +323,49 @@ define([
         return result;
     };
 
+    function getHDRShaderProgram(context, shaderProgram) {
```
Here and below, I _think_ the name would be `getHdrShaderProgram`.

pjcozzi commented on Source/Scene/GlobeDepth.js:
```diff
@@ -239,13 +241,15 @@ define([
         updateCopyCommands(this, context, width, height, passState);
         context.uniformState.globeDepthTexture = undefined;
+
+        this._useHDR = hdr;
```
Same comment.

And throughout.

pjcozzi commented on Source/Scene/Model.js:
```diff
@@ -2083,6 +2096,17 @@ define([
+                '\n' +
+                'void main() { \n' +
+                '    non_gamma_corrected_main(); \n' +
+                '    gl_FragColor.rgb = czm_gammaCorrect(gl_FragColor.rgb); \n' +
```
@bagnell @lilleyse @emackey on a related note - does this fix any of the gamma correction issues that I believe the glTF community has brought up about differences between Cesium and other WebGL engines when rendering glTF models?

pjcozzi commented on Source/Scene/PostProcessStageLibrary.js:
```diff
@@ -685,6 +695,88 @@ define([
+     * @return {PostProcessStage} A post-process stage that applies ACES tonemapping operator.
+     * @private
+     */
+    PostProcessStageLibrary.createACESTonemappingStage = function(useAutoExposure) {
```
Same naming comment.  Throughout.

pjcozzi commented on Source/Scene/PostProcessStageLibrary.js:
```diff
@@ -685,6 +695,88 @@ define([
     };
 
+    /**
+     * Creates a post-process stage that applies ACES tonemapping operator.
```
What is ACES?  And provide a tad of context please.

pjcozzi replied:
Same comment for the other tonemapping stages.  Provide enough context so users would know when to select one over the other.  Could make a nice Sandcastle example as well.

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -1405,6 +1411,55 @@ define([
+         * The value used for gamma correction.
+         * @memberof Scene.prototype
+         * @type {Number}
+         * @default 2.4
```
Above it is set to `2.2`

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -1405,6 +1411,55 @@ define([
+            },
+            set : function(value) {
+                var context = this._context;
+                var hdr = value && context.depthTexture && (context.colorBufferFloat || context.colorBufferHalfFloat);
```
Here, especially below, and throughout, these statements are easier to read when written as `var foo = (expression);`.

Could be a good one for the Coding Guide if it isn't already in there.

pjcozzi commented on Source/Scene/Scene.js:
```diff
@@ -1405,6 +1411,55 @@ define([
+         *
+         * @memberof Scene.prototype
+         * @type {Color}
+         * @default Color(0.8,0.8,0.8)
```
Whitespace.

pjcozzi commented on Source/Scene/Tonemapper.js:
```diff
@@ -0,0 +1,58 @@
+    'use strict';
+
+    /**
+     * A tonemapping algorithm when rendering with high dynamic range.
```
Context throughout.

pjcozzi commented on Source/Shaders/GlobeFS.glsl:
```diff
@@ -235,13 +241,28 @@ void main()
     vec4 finalColor = color;
 #endif
 
+
```
Extra vertical whitespace?

pjcozzi commented on Source/Shaders/PostProcessStages/FilmicTonemapping.glsl:
```diff
@@ -0,0 +1,35 @@
+	vec3 c = ((color * (A * color + C * B) + D * E) / (color * ( A * color + B) + D * F)) - E / F;
+	float w = ((white * (A * white + C * B) + D * E) / (white * ( A * white + B) + D * F)) - E / F;
+
+    c = czm_inverseGamma(c / w);
```
Indentation.

pjcozzi commented on Source/Shaders/PostProcessStages/ACESTonemapping.glsl:
```diff
@@ -0,0 +1,31 @@
+#ifdef AUTO_EXPOSURE
+    color /= texture2D(autoExposure, vec2(0.5)).r;
+#endif
+    float g = 0.985;
```
A thought for when auto exposure is not used here and potential for other tone mappers: is it possible to select good constants based on indoor vs. outdoor, global vs. local, satellite imagery vs. vectors, photogrammetry vs. lots of primitives, etc. - could we provide a handful of decent defaults to select from?  Or does it just vary too widely based on the frequency of whatever content an app loads?

Also, what is auto exposure on by default?  What is wrong with it and what is the path to correct it now or later?

pjcozzi commented:
I know we are starting to plan IBL, but can we also push this one over the finish line before we start coding for IBL unless we think we need to knowledge from IBL to help this one.

lilleyse commented:
## A bunch of side-by-sides between master (left) and hdr (right).

##### It's a little harder to see ground primitives in some cases. The red ground primitive is fine, just looks slightly different.

![corridor](https://user-images.githubusercontent.com/915398/46630624-c1701c80-cb12-11e8-88d7-9f87bbde1350.jpg)
![classification](https://user-images.githubusercontent.com/915398/46630646-c2a14980-cb12-11e8-84f0-98c7fcb20433.jpg)

##### These images are both from the `hdr` branch, but showing night and day. It's hard to tell if there is any shading. The globe shading may need to be stronger when hdr is enabled.

![enable-lighting](https://user-images.githubusercontent.com/915398/46630625-c1701c80-cb12-11e8-8111-6cffa3fbafc5.jpg)

##### Some geometries appear faded compared to `master`. A lot of other geometries seemed fine so maybe this is just affecting batched primitives?

![geojson](https://user-images.githubusercontent.com/915398/46630626-c1701c80-cb12-11e8-8b86-5a3c633a8df0.jpg)
![hide-show-geometries](https://user-images.githubusercontent.com/915398/46630629-c1701c80-cb12-11e8-83aa-c9b4091c9262.jpg)
![kmz](https://user-images.githubusercontent.com/915398/46630630-c1701c80-cb12-11e8-8abd-02be6c85b601.jpg)

##### The globe and atmosphere colors seem to lose some vibrancy. This might be a matter of tweaking the ground atmosphere to look better for HDR.

![globe](https://user-images.githubusercontent.com/915398/46630627-c1701c80-cb12-11e8-91be-f92e70dfbcf9.jpg)
![globe-atmosphere](https://user-images.githubusercontent.com/915398/46630628-c1701c80-cb12-11e8-878c-5f9409227aaf.jpg)
![atmosphere](https://user-images.githubusercontent.com/915398/46630642-c2a14980-cb12-11e8-987a-b4483ec5d985.jpg)

##### That includes lowering the night time fade out distance because the change in the lit/dark side is less noticeable.

![sunrise](https://user-images.githubusercontent.com/915398/46630640-c2a14980-cb12-11e8-9d5e-f1f1501cdbe9.gif)

##### Some labels look different in the `hdr` branch:
![label-background-color](https://user-images.githubusercontent.com/915398/46631225-d057ce80-cb14-11e8-8ee9-95b6463de6a7.jpg)
![label-text](https://user-images.githubusercontent.com/915398/46631228-d2219200-cb14-11e8-8ee8-dfb0b8f6fc55.jpg)

##### glTF 1.0 models often look washed out
![plane](https://user-images.githubusercontent.com/915398/46630637-c208b300-cb12-11e8-84a0-4552d87522ea.jpg)
![3d-tiles-interior](https://user-images.githubusercontent.com/915398/46630641-c2a14980-cb12-11e8-9f20-32cf5c799de9.jpg)
![balloons](https://user-images.githubusercontent.com/915398/46630644-c2a14980-cb12-11e8-9349-1b6b195e9c6a.jpg)
![bim](https://user-images.githubusercontent.com/915398/46630645-c2a14980-cb12-11e8-8d79-d369faec660c.jpg)

##### I'm also seeing some washed out point clouds

![point-cloud](https://user-images.githubusercontent.com/915398/46630638-c208b300-cb12-11e8-98ed-10b9846442ce.jpg)
![point-cloud-style](https://user-images.githubusercontent.com/915398/46630639-c2a14980-cb12-11e8-9deb-be691108f621.jpg)

##### The scene's background color looks too different in the `hdr` branch:

![background-color](https://user-images.githubusercontent.com/915398/46630643-c2a14980-cb12-11e8-9340-9f2717c67c7f.jpg)

##### In general imagery looks good with `hdr`. One example is the Natural Earth imagery.

![natural-earth](https://user-images.githubusercontent.com/915398/46630635-c208b300-cb12-11e8-9ef5-078a2b8c0863.jpg)

##### The composite material in the `Materials` demo looks faded in `hdr`

![composite-material](https://user-images.githubusercontent.com/915398/46630647-c2a14980-cb12-11e8-9da6-865b634d7943.jpg)

##### Imagery contrast adjustment looks too strong

![contrast](https://user-images.githubusercontent.com/915398/46630648-c2a14980-cb12-11e8-939b-0f98bc1a5b1a.jpg)

lilleyse commented:
Some bugs I noticed:

* Shadows aren't working (or aren't visible)
* `Clamp to 3D Model` demo is not clamping when paused
* `Clamp to 3D Tiles demo` seems to be ignoring `objectsToExclude`
* `Depth of Field` might not be working, the blur is too strong
* `Globe Materials` crashes
* Picking and drill picking seem to be broken

I wouldn't be surprised if a few of these are related to the recent Scene cleanup changes.

lilleyse commented on Source/Scene/FrameState.js:
```diff
@@ -317,6 +317,13 @@ define([
+         *
+         * @type {Color}
+         */
+        this.sunColor = undefined;
```
I wonder if this should tie into https://github.com/AnalyticalGraphicsInc/cesium/pull/7025

lilleyse commented on Source/Scene/Model.js:
```diff
@@ -1936,6 +1938,17 @@ define([
             drawFS = 'uniform vec4 czm_pickColor;\n' + drawFS;
         }
 
+        if (model._upgrade10To20) {
```
The version detection changed a bit with recent Model update. The source version should now be stored in `gltf.extras.sourceVersion`. Resolving this should fix the glTF 1 screenshots in https://github.com/AnalyticalGraphicsInc/cesium/pull/7017#issuecomment-427966855.

lilleyse commented on Source/Scene/PostProcessStageCollection.js:
```diff
@@ -59,14 +61,27 @@ define([
 
+        // Auto-exposure is currently disabled because most shaders output a value in [0.0, 1.0].
+        // Some shaders, such as the atmosphere and ground atmosphere, output values slightly over 1.0.
+        this._autoExposureEnabled = false;
```
Instead could the stage be disabled with `this._autoExposure.enabled = false`?

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -1444,6 +1450,55 @@ define([
         },
 
+        /**
+         * The value used for gamma correction.
```
Maybe it should be noted that this only applied when hdr is on.

lilleyse commented on Source/Scene/Scene.js:
```diff
@@ -1444,6 +1450,55 @@ define([
+         *
+         * @memberof Scene.prototype
+         * @type {Color}
+         * @default Color(0.8, 0.8, 0.8)
```
Update the default to `(0.8, 0.85, 1.0)`

bagnell replied:
Updated. @ggetz are we sure we don't want to use this value? https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/processPbrMaterials.js#L620

bagnell commented:
> Now that I've looked at the code most of the screenshots in [#7017 (comment)](https://github.com/AnalyticalGraphicsInc/cesium/pull/7017#issuecomment-427966855) are probably just missing gamma correction in some area or need slight tweaks.

Yes, most were missing gamma correction. I think I've addressed all of the comments where you posted images.

mramato commented on Source/Scene/PostProcessStageLibrary.js:
```diff
@@ -16,12 +17,16 @@ define([
         '../Shaders/PostProcessStages/LensFlare',
+        '../Shaders/PostProcessStages/ModifiedReinhardTonemapping',
         '../Shaders/PostProcessStages/NightVision',
+        '../shaders/PostProcessStages/ReinhardTonemapping',
```
lowercase shaders here is breaking linux.

bagnell replied:
Thanks. Fixed.

bagnell commented on Source/Scene/AutoExposure.js:
```diff
@@ -0,0 +1,432 @@
+            wrapS : TextureWrap.CLAMP_TO_EDGE,
+            wrapT : TextureWrap.CLAMP_TO_EDGE,
+            minificationFilter : TextureMinificationFilter.NEAREST,
+            magnificationFilter : TextureMagnificationFilter.NEAREST
```
Probably. Where would you sample? Between two pixels? That would also impact the samples count.

Since this isn't used right now, I'd put that in future work. I'll update the list above.

bagnell commented on Source/Scene/AutoExposure.js:
```diff
@@ -0,0 +1,432 @@
+     * @private
+     */
+    AutoExposure.prototype.clear = function(context) {
+        var framebuffers = this._framebuffers;
```
No, but I think it's good practice. For example, The framebuffers we render to every frame, we render to the entire textures, but we still clear them. Ass far as performance, the slowest part of a clear call should be the driver call.

lilleyse commented:
Ran through the Sandcastle demos again. Mostly everything looks good now. These are the only things I noticed:

The label background color is lighter:

![hdr-1](https://user-images.githubusercontent.com/915398/46631225-d057ce80-cb14-11e8-8ee9-95b6463de6a7.jpg)

Point clouds look better now but the classification volumes look a bit faded:

![hdr-1](https://user-images.githubusercontent.com/915398/47532099-7594d580-d87d-11e8-90c3-2b379e1cd8a2.jpg)

The composite material in the Materials demo looks faded in hdr

![hdr-1](https://user-images.githubusercontent.com/915398/46630647-c2a14980-cb12-11e8-9da6-865b634d7943.jpg)

bagnell commented:
@lilleyse Updated.

lilleyse commented:
Thanks the recent updates look good. Hopefully we've caught all of the areas that need gamma correction. I'll merge soon after the release.

lilleyse commented:
Billboards seems to disappear when I toggle hdr off in the KML example. Is toggling hdr on and off sufficiently tested?

[Local example](http://localhost:8080/Apps/Sandcastle/index.html#c=zVVtb9owEP4rFl8SJGrY9mUatFqBbp36qqadtC3TZJwjsXCcyHZo6cR/3zkJNA1F27QvixAi53vunnvuziyZJksB96DJIVFwTyZgRJHSz6XN93j5OsmUZUKB9rrDUC0Rk+VWZMog6GeoCD6cpaAZeVdHo4aDAlpZexsXhJpdF2cN1RojhypgKuLMWAmURdFtlskZ0xegCv9bncjCg8UY3tnFOTkgH2U2Y5IEXIDiQD4wLqSwAoxXJ0WSIIE7yLxQ3LH2uxvS7qnJVEzpXK5OsxT8gSu05RIxy4Ks0KiJI+fXUp2lcro9oTJjke9R2sdPwNJcgjvsL1LZn2/JNX5SPPF6Gz27m7RrFKRHdkr+Su6FTQikM4giiIij5FSYXpMcW8hZLiz7D0uPo/wa9KSk93oweItVP/5p1fEDcb3GN1CmHLoDMhYLIDcigr+t9V8LmWFil7fdNmoTUP429VOaboNDU3OZ8QU1SVbI6FiJlFnAXZozaWDYAuC2aczoFvQpLAVlq/mJwY5Xn5CvxQOvO3wxHfb2NvPLMG2qbYINlNWMLyA6calWmL7ED/e6Vx1o+L8UZj+8kiQtpBW5FGW9bwa/c28raHXRFnDd1GTdHLXvPYKqlVdMKdyz60eDAet6sjNOL4yShjRbwrGU/ib+M5bl9w1TsaNYj9xka6N3l+Oru8vpyXQvNrCQt6DORIMvwe3JxY/J+dXkbLjvDo1jCePCWqzBO53e4Ng6lXpPlfEEXIPaBVb3cyLiZLpSLBV8U0Ht7xJ2h51eZ2TsSsJRBX4v0jzTlhRa+rg/FnB/sDemPysQZCk3xmk06m9Ao0gsiYgOw07rzybsEC6ZMXgyL6QMxCOEnaNRH/2fwdy6ChVfLUFLtnIuyauj88pIKR318XUXVTe+EfEX)

bagnell commented:
> Billboards seems to disappear when I toggle hdr off in the KML example.

For the KML example, `Sandcastle.reset` is removing the data source. If you comment out that line and toggle HDR, it'll be fine.

> Is toggling hdr on and off sufficiently tested?

I haven't tested it exhaustively for all of the primitives, but I shouldn't need to since all it does is switch the derived command executed. If it works for one, it should work for all of them.

bagnell commented:
> Sorry to hold this up with more side by sides, but I think it's worth improving how ground atmosphere looks before we merge.

There's not much I can do about that. It's the result of tonemapping. I like the way it looks since you can get views like at the bottom of this post: https://cesium.com/blog/2018/08/23/siggraph-2018-trip-report/

If you think it should change, perhaps @ggetz could take a look.

lilleyse commented:
I added a bit of saturation and modified the darkness falloff of the globe. HDR vs. non-HDR still have a different "feel" but I'm happy enough with the way it looks now.

![hdr](https://user-images.githubusercontent.com/915398/48165877-f65bc480-e2b4-11e8-97f4-750dc85ac3eb.gif)

lilleyse commented:
> For the KML example, Sandcastle.reset is removing the data source. If you comment out that line and toggle HDR, it'll be fine.

Ahh that's my bad.

lilleyse commented:
Did a final sweep and noticed just a couple issues:

The power plant tileset was updated to be glTF2 + `KHR_techniques_webgl` and doesn't seem to be gamma corrected properly anymore.
<img width="293" alt="powerplant" src="https://user-images.githubusercontent.com/915398/48209111-5b113080-e342-11e8-9ee4-764fccb125a6.PNG">

I'm seeing a crash when loading any demo in IE.
<img width="466" alt="ie-crash" src="https://user-images.githubusercontent.com/915398/48209282-b93e1380-e342-11e8-85ca-2be8f12346ed.PNG">

Firefox, Edge, and Chrome all work fine. I haven't tried Safari.

bagnell commented:
@lilleyse Updated.

lilleyse commented:
I think it would be better to apply the gamma correction in the Model layer since https://github.com/AnalyticalGraphicsInc/cesium/pull/7017/commits/4c33a68d12e034603eb209d5d4f8c989f7e29e83 will only work for 3D Tiles. In the same way that we detect a source model is gltf 1.0 [here](https://github.com/AnalyticalGraphicsInc/cesium/blob/hdr/Source/Scene/Model.js#L2012-L2021) can we detect if the source model uses `KHR_techniques_webgl`?

bagnell commented:
@lilleyse Yes, thanks. Updated.

lilleyse commented:
@bagnell this breaks glTF 2.0 PBR models because even they use `KHR_techniques_webgl` after PBR shaders are generated. The model will need to store whether it originally used the extension, like what `sourceVersion` does.

pjcozzi commented:
@bagnell as we discussed offline, please test performance - given the increased memory bandwidth - and suggest if this should default to true in general and for mobile in particular - or if we should default to false and/or encourage mobile to turn it off.

bagnell commented:
@pjcozzi There are two fullscreen framebuffers that increased in size. If the half-float extension is available, they'll be 2x as big. If not and float textures are supported, they'll be 4x as big. I didn't see a performance difference on a Pixel. On the iPad and iPhone, HDR wasn't enabled by default.

@lilleyse Updated. Is there anything else?

lilleyse commented:
Thanks @bagnell, that model issue was the last thing.

lilleyse commented:
Note for everyone: if you see any colors that look more faded than before make sure to open an issue. This usually means something is missing gamma correction. See some of the [side-by-sides](https://github.com/AnalyticalGraphicsInc/cesium/pull/7017#issuecomment-427966855) for examples of what this would look like.

We went through all the demos pretty carefully but it's possible there are areas we missed.
